test/test_torch.py

```diff
@@ -6933,6 +6933,82 @@ def test_bitwise_xor(self, device):
                          torch.bitwise_xor(torch.tensor([True, True, False], device=device),
                                            torch.tensor([False, True, False], device=device)))
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    @dtypes(*list(product(torch.testing.get_all_dtypes(include_complex=False),
+                          torch.testing.get_all_dtypes(include_complex=False))))
+    def test_heaviside(self, device, dtypes):
```

Comment:
This is a pretty cool test but I think you can simplify it by generating an array with NumPy's [integers](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.integers.html?highlight=integers#numpy.random.Generator.integers) and then a second NumPy array with integers or rand depending on the dtype of the rhs argument. Then you can run NumPy's heaviside on these args to get your expected values.

For your actual values you can `torch.from_numpy(lhs/rhs).to(device=device, dtype=dtype)` the NumPy arrays and run your `torch.heaviside`. Then you can compare the results with `self.assertEqual(actual, expected, exact_dtype=False)`, and verify the `torch.heaviside` function produced the correct dtype by checking the dtype of the `actual` tensor against the result of `torch.result_type`. 

This should let you write a thorough but compact test, I think. What are your thoughts?

Reply:
Wow, LGTM. I would improve my test case, thanks for your advice! 👍

Reply:
Tests look good. This first test will need a tweak depending on how you decide to handle type promotion. It's probably easiest to require the types be the same, then you could just check for that case and that a runtime error is thrown.